File: src/elixir/svd/test/extensions/service_provider_test.exs
defmodule SVD.Extensions.ServiceProviderTest do
  use ExUnit.Case, async: true

  alias SVD.Extensions.ServiceProvider
  alias SVD.TestSupport.Faker
  alias SVD.Validators.StandardisedVesselDatasetValidator

  defmodule CustomValidator do
    def validate(_), do: []
  end

  test "new service provider" do
    provider = ServiceProvider.new()

    assert provider != nil
    assert ServiceProvider.get_validator(provider) != nil
    assert ServiceProvider.get_json_exporter(provider) != nil
    assert ServiceProvider.get_csv_exporter(provider) != nil
    assert ServiceProvider.get_xml_exporter(provider) != nil
  end

  test "new service provider with custom validator" do
    provider = ServiceProvider.new_with_custom_validator(CustomValidator)

    assert provider != nil
    assert ServiceProvider.get_validator(provider) == CustomValidator
  end

  test "set validator" do
    provider = ServiceProvider.new()
    original = ServiceProvider.get_validator(provider)

    updated = ServiceProvider.set_validator(provider, CustomValidator)

    assert original != ServiceProvider.get_validator(updated)
    assert ServiceProvider.get_validator(updated) == CustomValidator
  end

  test "set specific exporters" do
    provider = ServiceProvider.new()

    json = SVD.Exporters.Formats.JSONExporter.new(CustomValidator)
    csv = SVD.Exporters.Formats.CSVExporter.new(CustomValidator)
    xml = SVD.Exporters.Formats.XMLExporter.new(CustomValidator)

    provider =
      provider
      |> ServiceProvider.set_json_exporter(json)
      |> ServiceProvider.set_csv_exporter(csv)
      |> ServiceProvider.set_xml_exporter(xml)

    assert ServiceProvider.get_json_exporter(provider) == json
    assert ServiceProvider.get_csv_exporter(provider) == csv
    assert ServiceProvider.get_xml_exporter(provider) == xml
  end

  test "export with json exporter" do
    provider = ServiceProvider.new()
    exporter = ServiceProvider.get_json_exporter(provider)

    assert {:ok, result} =
             SVD.Exporters.Formats.JSONExporter.export_async(exporter, Faker.valid_svd())

    assert byte_size(result.data) > 0
  end

  test "export with csv exporter" do
    provider = ServiceProvider.new()
    exporter = ServiceProvider.get_csv_exporter(provider)

    assert {:ok, result} =
             SVD.Exporters.Formats.CSVExporter.export_async(exporter, Faker.valid_svd())

    assert byte_size(result.data) > 0
  end

  test "export with xml exporter" do
    provider = ServiceProvider.new()
    exporter = ServiceProvider.get_xml_exporter(provider)

    assert {:ok, result} =
             SVD.Exporters.Formats.XMLExporter.export_async(exporter, Faker.valid_svd())

    assert byte_size(result.data) > 0
  end

  test "validate with validator" do
    provider = ServiceProvider.new()
    validator = ServiceProvider.get_validator(provider)

    assert validator.validate(Faker.valid_svd()) == []
    assert validator.validate(Faker.invalid_svd()) != []
  end

  test "all exporters use provided validator" do
    provider = ServiceProvider.new_with_custom_validator(StandardisedVesselDatasetValidator)
    invalid = Faker.invalid_svd()

    assert {:error, _} =
             provider
             |> ServiceProvider.get_json_exporter()
             |> SVD.Exporters.Formats.JSONExporter.export_async(invalid)

    assert {:error, _} =
             provider
             |> ServiceProvider.get_csv_exporter()
             |> SVD.Exporters.Formats.CSVExporter.export_async(invalid)

    assert {:error, _} =
             provider
             |> ServiceProvider.get_xml_exporter()
             |> SVD.Exporters.Formats.XMLExporter.export_async(invalid)
  end
end

File: src/go/svd/extensions/service_provider_test.go
package extensions_test

import (
	"context"
	"errors"
	"testing"

	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/exporters"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/extensions"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/internal/testdata"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/models"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/validators"
)

type alwaysInvalidValidator struct{}

func (alwaysInvalidValidator) Validate(*models.StandardisedVesselDataset) []validators.ValidationError {
	return []validators.ValidationError{{Field: "General.Imo", Message: "Imo is required"}}
}

type staticExporter struct {
	result *exporters.StandardisedVesselDatasetContent
	err    error
}

func (s staticExporter) ExportAsync(context.Context, *models.StandardisedVesselDataset) (*exporters.StandardisedVesselDatasetContent, error) {
	return s.result, s.err
}

func TestServiceProvider(t *testing.T) {
	t.Run("default_provider", func(t *testing.T) {
		p := extensions.NewServiceProvider()
		if p.GetValidator() == nil || p.GetJSONExporter() == nil || p.GetCSVExporter() == nil || p.GetXMLExporter() == nil {
			t.Fatal("provider returned nil services")
		}
	})

	t.Run("custom_validator", func(t *testing.T) {
		p := extensions.NewServiceProviderWithCustomValidator(alwaysInvalidValidator{})
		_, err := p.GetJSONExporter().ExportAsync(context.Background(), testdata.ValidSVD())
		if err == nil {
			t.Fatal("expected validation error")
		}
	})

	t.Run("set_validator_updates_exporters", func(t *testing.T) {
		p := extensions.NewServiceProvider()
		p.SetValidator(alwaysInvalidValidator{})
		_, err := p.GetCSVExporter().ExportAsync(context.Background(), testdata.ValidSVD())
		if err == nil {
			t.Fatal("expected validator error")
		}
	})

	t.Run("set_custom_exporters", func(t *testing.T) {
		p := extensions.NewServiceProvider()
		customResult := &exporters.StandardisedVesselDatasetContent{Data: []byte("x"), FileName: "x", ContentType: "text/plain"}
		customErr := errors.New("custom")

		p.SetJSONExporter(staticExporter{result: customResult})
		p.SetCSVExporter(staticExporter{err: customErr})
		p.SetXMLExporter(staticExporter{result: customResult})

		if got, err := p.GetJSONExporter().ExportAsync(context.Background(), nil); err != nil || got != customResult {
			t.Fatalf("unexpected json exporter result: %v %v", got, err)
		}
		if _, err := p.GetCSVExporter().ExportAsync(context.Background(), nil); !errors.Is(err, customErr) {
			t.Fatalf("unexpected csv exporter error: %v", err)
		}
		if got, err := p.GetXMLExporter().ExportAsync(context.Background(), nil); err != nil || got != customResult {
			t.Fatalf("unexpected xml exporter result: %v %v", got, err)
		}
	})
}
